scripts/scanTs.js

```diff
@@ -0,0 +1,50 @@
+const fs = require('fs');
+const path = require('path');
+const { Project, CallExpression } = require('ts-morph');
+
+const SRC_DIR = path.join(__dirname, '..', 'src');
+const project = new Project({
+  tsConfigFilePath: path.join(__dirname, '..', 'tsconfig.json'),
+});
+
+let detected = false;
+const scanDirectory = (dir) => {
+  const files = fs.readdirSync(dir);
+  files.forEach((file) => {
+    const fullPath = path.join(dir, file);
+    const stat = fs.statSync(fullPath);
+    if (stat.isDirectory()) {
+      scanDirectory(fullPath);
+    } else if (fullPath.endsWith('.ts')) {
+      analyzeFile(fullPath);
+    }
+  });
+};
+
+// This function will detect all the usages of fs.read* and send warnings with the location of the usage
+const analyzeFile = (filePath) => {
+  const srcFile = project.addSourceFileAtPath(filePath);
+  const funcCalls = srcFile.getDescendantsOfKind(CallExpression);
+
+  funcCalls.forEach((callExpression) => {
+    const exp = callExpression.getExpression();
+    if (exp.getText().startsWith('fs.read')) {
+      detected = true;
+      console.warn(
+        `Warning: Usage of "${exp.getText()}" in file "${filePath}" at line ${callExpression.getStartLineNumber()}.\n`
+      );
+    }
+  });
+};
+
+scanDirectory(SRC_DIR);
+
+if (detected) {
+  console.log('The warnings above do not mean the usages are wrong.');
+  console.log(`Avoid reading local artifacts with "fs.read*" since esbuild cannot bundle the artifacts together.`);
+  console.log('Consider using import instead or reach out to IDEx Foundations team');
```

Comment:
does this need to throw an error?

Reply:
No need. The usage can work when the target file is outside the project. But we cannot tell whether it is outside or inside. So we only send warning to developers to make them aware in case it blocks the development.